File: solutions/haskell/010-rotting-oranges.hs
{-
  Problem 10: Rotting Oranges (LeetCode 994)
  Difficulty: Med
  Language: Haskell
-}
import qualified Data.Set as Set

orangesRotting :: [[Int]] -> Int
orangesRotting grid = bfs rotten fresh 0
  where
    rows = length grid
    cols = length (head grid)
    indexed = [ ((r, c), grid !! r !! c) | r <- [0..rows-1], c <- [0..cols-1] ]
    rotten = Set.fromList [ pos | (pos, 2) <- indexed ]
    fresh  = Set.fromList [ pos | (pos, 1) <- indexed ]
    neighbors (r, c) = filter valid [(r-1,c),(r+1,c),(r,c-1),(r,c+1)]
    valid (r, c) = r >= 0 && r < rows && c >= 0 && c < cols

    bfs rot fr time
      | Set.null fr = time
      | Set.null newRotten = -1
      | otherwise = bfs (Set.union rot newRotten) (Set.difference fr newRotten) (time + 1)
      where
        newRotten = Set.fromList
          [ n | r <- Set.toList rot, n <- neighbors r, Set.member n fr ]

File: solutions/scala/010-rotting-oranges.scala
/**
 * Problem 10: Rotting Oranges (LeetCode 994)
 * Difficulty: Med
 * Language: Scala
 */
def orangesRotting(grid: Array[Array[Int]]): Int = {
  val rows = grid.length; val cols = grid(0).length
  val indexed = for (r <- 0 until rows; c <- 0 until cols) yield ((r, c), grid(r)(c))
  val rotten = indexed.collect { case ((r, c), 2) => (r, c) }.toList
  val fresh  = indexed.collect { case ((r, c), 1) => (r, c) }.toSet

  @annotation.tailrec
  def bfs(queue: List[(Int, Int)], fr: Set[(Int, Int)], time: Int): Int = {
    if (fr.isEmpty) time
    else if (queue.isEmpty) -1
    else {
      val (nextQ, nextFr) = queue.foldLeft((List.empty[(Int,Int)], fr)) {
        case ((nq, f), (r, c)) =>
          List((r-1,c),(r+1,c),(r,c-1),(r,c+1))
            .filter { case (nr,nc) => nr>=0 && nr<rows && nc>=0 && nc<cols && f.contains((nr,nc)) }
            .foldLeft((nq, f)) { case ((nq2, f2), pos) => (pos :: nq2, f2 - pos) }
      }
      if (nextQ.isEmpty) (if (nextFr.isEmpty) time else -1)
      else bfs(nextQ, nextFr, time + 1)
    }
  }
  bfs(rotten, fresh, 0)
}
